airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/operation/BigQueryStorageOperation.kt

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.destination.bigquery.operation
+
+import com.google.cloud.bigquery.BigQuery
+import com.google.cloud.bigquery.BigQueryException
+import com.google.cloud.bigquery.TableId
+import io.airbyte.cdk.integrations.util.ConnectorExceptionUtil
+import io.airbyte.commons.exceptions.ConfigErrorException
+import io.airbyte.integrations.base.destination.operation.StorageOperation
+import io.airbyte.integrations.base.destination.typing_deduping.StreamConfig
+import io.airbyte.integrations.base.destination.typing_deduping.StreamId
+import io.airbyte.integrations.base.destination.typing_deduping.TyperDeduperUtil
+import io.airbyte.integrations.destination.bigquery.BigQueryUtils
+import io.airbyte.integrations.destination.bigquery.formatter.BigQueryRecordFormatter
+import io.airbyte.integrations.destination.bigquery.typing_deduping.BigQueryDestinationHandler
+import io.airbyte.integrations.destination.bigquery.typing_deduping.BigQuerySqlGenerator
+import io.airbyte.protocol.models.v0.DestinationSyncMode
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.time.Instant
+import java.util.*
+
+private val log = KotlinLogging.logger {}
+
+abstract class BigQueryStorageOperation<Data>(
+    protected val bigquery: BigQuery,
+    private val sqlGenerator: BigQuerySqlGenerator,
+    private val destinationHandler: BigQueryDestinationHandler,
+    protected val datasetLocation: String
+) : StorageOperation<Data> {
+    private val existingSchemas = HashSet<String>()
+    override fun prepareStage(streamId: StreamId, destinationSyncMode: DestinationSyncMode) {
+        // Prepare staging schema
+        createStagingSchema(streamId)
+        // Prepare staging table. For overwrite, it does drop-create so we can skip explicit create.
+        if (destinationSyncMode == DestinationSyncMode.OVERWRITE) {
+            truncateStagingTable(streamId)
+        } else {
+            createStagingTable(streamId)
+        }
+    }
+
+    private fun createStagingSchema(streamId: StreamId) {
```

Comment:
nit: createStagingDataset? (i.e. use bigquery terminology since this is bigquery-specific code)